cpu/stm32f1: add unified rtt configuration

fjmolinas opened the pull request:
### Contribution description

This PR is split from https://github.com/RIOT-OS/RIOT/pull/13874. It unifies `rtt` configuration for `stm32f1` and make the `RTT_FREQYENCY` configurable.

As in #13874 max and min values for the RTT are documented.

### Testing procedure

- `tests/periph_rtt` should still pass on `stm32f1` boards with different frequencies


`BOARD=iotlab-m3 make -C tests/periph_rtt flash test --no-print-directory -j3`
```
main(): This is RIOT! (Version: 2020.07-devel-56-g42700-pr_rtt_enhancement)

RIOT RTT low-level driver test
This test will display 'Hello' every 5 seconds

Initializing the RTT driver
RTT now: 0
Setting initial alarm to now + 5 s (81920)
Done setting up the RTT, wait for many Hellos
Hello
Hello
Hello
Hello
Hello
```

`CFLAGS+=-DRTT_FREQUENCY=1 BOARD=iotlab-m3 make -C tests/periph_rtt clean flash test --no-print-directory -j3`

```
main(): This is RIOT! (Version: 2020.07-devel-56-g42700-pr_rtt_enhancement)

RIOT RTT low-level driver test
This test will display 'Hello' every 5 seconds

Initializing the RTT driver
RTT now: 0
Setting initial alarm to now + 5 s (5)
Done setting up the RTT, wait for many Hellos
Hello
Hello
Hello
Hello
Hello
```

### Issues/PRs references

Split from https://github.com/RIOT-OS/RIOT/pull/13874

benpicco commented on boards/common/iotlab/include/periph_conf_common.h:
```diff
@@ -149,20 +149,6 @@ static const uart_conf_t uart_config[] = {
-#define RTT_IRQ             RTC_IRQn
-#define RTT_ISR             isr_rtc
-#define RTT_MAX_VALUE       (0xffffffff)
-#define RTT_FREQUENCY       (1)             /* in Hz */
```
You probably want to keep this line to retain the old behavior.

fjmolinas replied:
I was actually hopping to change the default to the highest frequency as well, but I could do that in a different PR.

benpicco commented on boards/fox/include/periph_conf.h:
```diff
@@ -149,20 +149,6 @@ static const spi_conf_t spi_config[] = {
-#define RTT_IRQ             RTC_IRQn
-#define RTT_ISR             isr_rtc
-#define RTT_MAX_VALUE       (0xffffffff)
-#define RTT_FREQUENCY       (1)             /* in Hz */
```
dito

fjmolinas commented:
Re-worked to leave only `RTT_FREQUENCY` definition in `periph_conf.h` since it is a per board configuration. I kept the same configurations as before, but you can eventually override the default `RTT_FREQUENCY` at application level.

fjmolinas commented:
Seems there is a naming conflict..

fjmolinas commented:
> Seems there is a naming conflict..

Oh no, its because those `BOARD`s do not have rtt feature but are still defining `RTT_MAX_VALUE` with this change.

fjmolinas commented:
> Oh no, its because those BOARDs do not have rtt feature but are still defining RTT_MAX_VALUE with this change.

Seem like they both have a LSE oscilator. But I'll guard the definition against `MODULE_PERIPH_RTT` anyway.

fjmolinas commented:
> Seem like they both have a LSE oscilator. But I'll guard the definition against MODULE_PERIPH_RTT anyway.

And add rtt in a follow-up.

fjmolinas commented:
> And add rtt in a follow-up.

Actually, does it make sense enabling right away? the bluepill config already has `CLOCK_LSE` set to 1, and `periph_rtc` is already provideed by default for all stm32f1.

fjmolinas commented:
> Actually, does it make sense enabling right away? the bluepill config already has CLOCK_LSE set to 1, and periph_rtc is already provideed by default for all stm32f1.

I wouldn't have any board to test though.

fjmolinas commented:
@MrKevinWeiss do you have some blxxxpill?

jue89 commented:
> > Actually, does it make sense enabling right away? the bluepill config already has CLOCK_LSE set to 1, and periph_rtc is already provideed by default for all stm32f1.
> 
> I wouldn't have any board to test though.

I actually had a PR in the pipe line enabling `periph_rtt` on the `blxxxpill`. I would happily test your changes on the `bluepill`.

fjmolinas commented:
> I actually had a PR in the pipe line enabling periph_rtt on the blxxxpill. I would happily test your changes on the bluepill.

I added `RTT_FREQUENCY` for the bluepill and enabled `periph_rtt` for all stm32f1, just as for `periph_rtc`, in that commit I removed `FEATURES_PROVIDED += periph_rtt` from `stm32f1`  boards, since its provided by the cpu.

fjmolinas commented:
> I added RTT_FREQUENCY for the bluepill and enabled periph_rtt for all stm32f1, just as for periph_rtc, in that commit I removed FEATURES_PROVIDED += periph_rtt from stm32f1 boards, since its provided by the cpu.

Seem that with the github issues, my comment got lost. I wanted to provide `periph_rtt` at cpu level, but then realized that at least 3 stm32f1 boards do not have `CLOCK_LSE`, so I just kept the original commit and split the `blxxx`change into #13929.

fjmolinas commented:
Seems github issues got in the way here.

```
-- github reports HEAD of master as 341916cd307bb9fcacf8a457f1d5988528c6ca5c
fatal: unable to access 'https://github.com/RIOT-OS/RIOT.git/': The requested URL returned error: 504
--- creating merge commit ...
-- merging c0d1884813e1816b5cf445b21464fa5d7134da9f into 341916cd307bb9fcacf8a457f1d5988528c6ca5c
--- cloning base repo
git-cache: cloning from cache. tag=commit341916cd307bb9fcacf8a457f1d5988528c6ca5c
--- adding remotes
--- checking out merge branch
Switched to a new branch 'pull/341916cd307bb9fcacf8a457f1d5988528c6ca5c/c0d1884813e1816b5cf445b21464fa5d7134da9f'
--- fetching pr_stm32f1_rtt
remote: 
remote: 
remote: 
remote: 
502 Bad Gateway
```

fjmolinas commented:
Ajjj is starting to be more annoying than I though, because `periph_cpu` is included before `periph_conf` I can't have a default value there which is overridden per `BOARD`. Providing them only per `BOARD` is ok, but now `rtc_rtt` unit tests is complaining about `RTT_MAX_VALUE` missing, anyway this last one should make murdock happy, sorry for the force pushing.

aabadie commented on boards/common/iotlab/include/periph_conf_common.h:
```diff
@@ -149,19 +149,16 @@ static const uart_conf_t uart_config[] = {
 #define UART_NUMOF          ARRAY_SIZE(uart_config)
 /** @} */
 
+#if defined(MODULE_PERIPH_RTT) || defined(DOXYGEN)
```
I think adding this preprocessor statement here is not needed. Can you explain you added it ?
If there's no good reason, then it should be removed.

fjmolinas replied:
Because of this:

https://github.com/RIOT-OS/RIOT/blob/341916cd307bb9fcacf8a457f1d5988528c6ca5c/drivers/include/periph/rtt.h#L45-L56

Otherwise `RTT_FREQUENCY` would be defined but not `RTT_MAX_VALUE`... I could actually also remove the check and have those line do:

```
#ifndef RTT_FREQUENCY

/* Allow mock-RTT for unit tests */
#ifdef MOCK_RTT_FREQUENCY
#define RTT_FREQUENCY MOCK_RTT_FREQUENCY
#else

#warning "RTT_FREQUENCY undefined. Set RTT_FREQUENCY to the number of ticks" \
         "per second for the current architecture."
#endif
#endif

#ifndef RTT_MAX_VALUE 

/* Allow mock-RTT for unit tests */
#ifdef MOCK_RTT_FREQUENCY
#define RTT_MAX_VALUE MOCK_RTT_MAX_VALUE
#else

#warning "RTT_MAX_VALUE  undefined. Set RTT_MAX_VALUE  to the rtt timer size"
#endif
#endif
```

aabadie replied:
This is rtt.h so you are supposed to have the `periph_rtt` module already added to the build. So this is IMHO a non issue :)
I would remove all preprocessor check (even in cpu): I tried locally and it just builds fine.

fjmolinas replied:
> This is rtt.h so you are supposed to have the periph_rtt module already added to the build. So this is IMHO a non issue :)

 Not the case for `unittests`

aabadie replied:
Thinking of it a second time. Your suggestion for rtt.h would make sense. Now I'm wondering if it is necessary to add this in this PR ?

aabadie replied:
> Not the case for unittests

I could build and flash `tests-rtt-rtc` without issues.

aabadie replied:
Using this diff:
```diff
diff --git a/boards/common/iotlab/include/periph_conf_common.h b/boards/common/iotlab/include/periph_conf_common.h
index b9bf38a95..948cf325a 100644
--- a/boards/common/iotlab/include/periph_conf_common.h
+++ b/boards/common/iotlab/include/periph_conf_common.h
@@ -149,7 +149,6 @@ static const uart_conf_t uart_config[] = {
 #define UART_NUMOF          ARRAY_SIZE(uart_config)
 /** @} */
 
-#if defined(MODULE_PERIPH_RTT) || defined(DOXYGEN)
 /**
  * @name    Real time counter configuration
  * @{
@@ -158,7 +157,6 @@ static const uart_conf_t uart_config[] = {
 #define RTT_FREQUENCY       (1)             /* in Hz */
 #endif
 /** @} */
-#endif /* defined(MODULE_PERIPH_RTT) || defined(DOXYGEN) */
 
 /**
  * @name    I2C configuration
diff --git a/boards/nucleo-f103rb/include/periph_conf.h b/boards/nucleo-f103rb/include/periph_conf.h
index 14fddd0df..4a142b5d0 100644
--- a/boards/nucleo-f103rb/include/periph_conf.h
+++ b/boards/nucleo-f103rb/include/periph_conf.h
@@ -119,14 +119,12 @@ static const uart_conf_t uart_config[] = {
 #define UART_NUMOF          ARRAY_SIZE(uart_config)
 /** @} */
 
-#if defined(MODULE_PERIPH_RTT) || defined(DOXYGEN)
 /**
  * @name    Real time counter configuration
  * @{
  */
 #ifndef RTT_FREQUENCY
 #define RTT_FREQUENCY       (16384)      /* in Hz */
-#endif
 /** @} */
 #endif /* defined(MODULE_PERIPH_RTT) || defined(DOXYGEN) */
 
diff --git a/cpu/stm32f1/include/periph_cpu.h b/cpu/stm32f1/include/periph_cpu.h
index fc6438495..1ec3f7480 100644
--- a/cpu/stm32f1/include/periph_cpu.h
+++ b/cpu/stm32f1/include/periph_cpu.h
@@ -133,7 +133,6 @@ typedef struct {
     uint8_t chan;           /**< CPU ADC channel connected to the pin */
 } adc_conf_t;
 
-#if defined(MODULE_PERIPH_RTT) || defined(DOXYGEN)
 /**
  * @name    Real time counter configuration
  * @{
@@ -150,7 +149,6 @@ typedef struct {
 /* RTC frequency of 32kHz is not recommended, see RM0008 Rev 20, p490 */
 #define RTT_MAX_FREQUENCY   (RTT_CLOCK_FREQUENCY / 2) /* in Hz */
 /** @} */
-#endif /* defined(MODULE_PERIPH_RTT) || defined(DOXYGEN) */
 
 #ifdef __cplusplus
 }
```

fjmolinas replied:
Now try to build `bluepill`

aabadie replied:
<details><summary>test-rtt-rtc</summary>

```
make IOTLAB_NODE=auto-ssh BOARD=iotlab-m3 -C tests/unittests tests-rtt_rtc flash test
make: Entering directory '/work/riot/RIOT/tests/unittests'
Building application "tests_unittests" for "iotlab-m3" with MCU "stm32f1".

"make" -C /work/riot/RIOT/boards/iotlab-m3
"make" -C /work/riot/RIOT/boards/common/iotlab
"make" -C /work/riot/RIOT/core
"make" -C /work/riot/RIOT/cpu/stm32f1
"make" -C /work/riot/RIOT/cpu/cortexm_common
"make" -C /work/riot/RIOT/cpu/cortexm_common/periph
"make" -C /work/riot/RIOT/cpu/stm32_common
"make" -C /work/riot/RIOT/cpu/stm32_common/periph
"make" -C /work/riot/RIOT/cpu/stm32f1/periph
"make" -C /work/riot/RIOT/drivers
"make" -C /work/riot/RIOT/drivers/periph_common
"make" -C /work/riot/RIOT/drivers/rtt_rtc
"make" -C /work/riot/RIOT/sys
"make" -C /work/riot/RIOT/sys/embunit
"make" -C /work/riot/RIOT/sys/isrpipe
"make" -C /work/riot/RIOT/sys/newlib_syscalls_default
"make" -C /work/riot/RIOT/sys/pm_layered
"make" -C /work/riot/RIOT/sys/stdio_uart
"make" -C /work/riot/RIOT/sys/test_utils/interactive_sync
"make" -C /work/riot/RIOT/sys/tsrb
"make" -C /work/riot/RIOT/tests/unittests/tests-rtt_rtc
   text	   data	    bss	    dec	    hex	filename
  13232	    132	   2700	  16064	   3ec0	/work/riot/RIOT/tests/unittests/bin/iotlab-m3/tests_unittests.elf
iotlab-node --jmespath='keys(@)[0]' --format='int'  --list saclay,m3,12 --flash /work/riot/RIOT/tests/unittests/bin/iotlab-m3/tests_unittests.bin | grep 0
0
r
ssh -t abadie@saclay.iot-lab.info 'socat - tcp:m3-12.saclay.iot-lab.info:20000'
READY
s
START
....
OK (4 tests)

make: Leaving directory '/work/riot/RIOT/tests/unittests'
```

</details>

aabadie replied:
> Now try to build bluepill

Yes that's what I was testing :) and now I understand the issue: these boards don't provide the rtt feature and thus don't define RTT_FREQUENCY.

So I'm OK for modifying rtt.h like you suggested, this is the cleanest approach.

fjmolinas replied:
> Now try to build bluepill

Thats why I added the first `if defined(MODULE_PERIPH_RTT)` in `periph_cpu.h`, which required the second for `iotlab-m3` and other :(

fjmolinas replied:
> Thinking of it a second time. Your suggestion for rtt.h would make makes sense. Now I'm wondering if it is necessary to add this in this PR ?

Either Its done in a follow-up, and in that follow-up I remove the `defined(MODULE_PERIPH_RTT) || defined(DOXYGEN)`, or the other way around :)

fjmolinas replied:
> Thinking of it a second time. Your suggestion for rtt.h would make makes sense. Now I'm wondering if it is necessary to add this in this PR ?

Did it in #13933, if that one gets in fast we can avoid the header guards :) (I would prefer that)

aabadie replied:
#13933 is merged, please rebase and adapt this PR :)

fjmolinas replied:
yep on it

fjmolinas commented:
@benpicco does your ACK still stand?

fjmolinas commented:
Thanks for the review!
